dashboard/app/controllers/api/v1/user_school_infos_controller.rb

```diff
@@ -10,49 +10,36 @@ def update_last_confirmation_date
     head :no_content
   end
 
-  # PATCH /api/v1/users_school_infos/<id>/update_end_date
-  def update_end_date
-    ActiveRecord::Base.transaction do
-      @user_school_info.update!(end_date: DateTime.now, last_confirmation_date: DateTime.now)
-      @user_school_info.user.update!(properties: {last_seen_school_info_interstitial: DateTime.now})
-    end
-
-    head :no_content
-  end
-
-  # PATCH /api/v1/user_school_infos/<id>/update_school_info_id
-  # A new row is added to the user school infos table when a teacher updates current school to a different school,
-  # then the school_info_id is updated in the users table.
-
-  def update_school_info_id
-    ActiveRecord::Base.transaction do
-      school = @user_school_info.school_info.school
-      unless school
-        school = School.create!(new_school_params)
-        school.school_info.create!(school_info_params)
+  # PATCH /api/v1/users_school_infos/<id>/school_info
+  def update
+    existing_school_info = current_user.school_info
+    if school_info_params[:school_id].present?
+      new_school = SchoolInfo.find_or_create_by!(school_info_params.merge(validation_type: SchoolInfo::VALIDATION_NONE))
+      if existing_school_info == new_school && existing_school_info.complete?
+        current_user.user_school_infos.where(school_info: new_school).update(last_confirmation_date: DateTime.now)
+      else
+        current_user.update!(school_info_id: new_school.id)
+      end
+    # partial school info entered, so update it
+    elsif school_info_params[:country].present?
+      new_school_info = if existing_school_info && !existing_school_info.complete?
+                          SchoolInfo.find_or_create_by!(school_info_params.merge(validation_type: SchoolInfo::VALIDATION_NONE))
```

Comment:
What's the the difference between `find_or_create_by` and `find_or_create`?

Reply:
I can't find the `find_or_create` method.  Do you mean [`first_or_create`](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/first_or_create)?

There's a subtle but _very_ important distinction between these two methods. [`find_or_create_by`](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_create_by) finds the first record with the given attributes, or creates a record with the attributes if one is not found. Here's its implementation:

```rb
    def find_or_create_by(attributes, &block)
      find_by(attributes) || create(attributes, &block)
    end
```

On the other hand, `first_or_create` tries to get the first record of the relation, not the one matching given attributes. If no record is found at all, it creates one using the specified attributes. It's implemented like this:

```rb
    def first_or_create(attributes = nil, &block) # :nodoc:
      first || create(attributes, &block)
    end
```

There's [an excellent explanation here](https://apidock.com/rails/ActiveRecord/Relation/first_or_create#1651-RE-FALSE-Creates-record-by-given-attributes-only-if-table-is-empty) of how `first_or_create` is commonly misunderstood, and how it should be used.  I'll try to summarize:

```rb
# Supposed you wanted to find the user with email `brad@example.com`,
# or create it if it doesn't exist.

# You could
User.find_or_create_by(email: 'brad@example.com')
# which would do what you want.

# You might try
User.first_or_create(email: 'brad@example.com')
# but you might be surprised when the returned user has email "john@example.com"
# because it grabbed the first row in the table.

# The way to get what you want with first_or_create is with a pre-filtered relation:
User.where(email: 'brad@example.com').first_or_create

Reply:
Thank you for asking about this! We actually ended up with a _perfect_ situation to use `first_or_create` in our controller method.